### PR TITLE
[WIP] OAuth Improvement: Change user.id to use sub from IdP

### PR DESCRIPTION
**Pull request by mraible**
This PR is a work-in-progress. I still need to make everything work with OAuth + microservices. I'm creating this PR to see what fails on Travis.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [x] Tests are added where necessary
- [x] Documentation is added/updated where necessary
- [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by mraible** [comment]
I'm getting the following error when trying to start a gateway generated with this branch. Any ideas @jhipster/developers?

```
org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is java.lang.RuntimeException: 
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'servletEndpointRegistrar' defined in class path resource 
[org/springframework/boot/actuate/autoconfigure/endpoint/web/ServletEndpointManagementContextConfiguration.class]: Bean instantiation via factory method failed; nested exception is 
org.springframework.beans.BeanInstantiationException: Failed to instantiate 
[org.springframework.boot.actuate.endpoint.web.ServletEndpointRegistrar]: Factory method 'servletEndpointRegistrar' threw exception; nested exception is 
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 
'org.springframework.cloud.netflix.zuul.ZuulProxyAutoConfiguration$EndpointConfiguration': 
Post-processing of merged bean definition failed; nested exception is java.lang.IllegalStateException: Failed to introspect Class 
[org.springframework.cloud.netflix.zuul.ZuulProxyAutoConfiguration$EndpointConfiguration] from ClassLoader [sun.misc.Launcher$AppClassLoader@2a139a55]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:153)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:138)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:388)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:327)
	at com.mycompany.myapp.GatewayApp.main(GatewayApp.java:67)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49)
...
Caused by: java.lang.ClassNotFoundException: org.springframework.boot.actuate.trace.http.HttpTraceRepository
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 76 common frames omitted
```

**Comment by deepu105** [comment]
@mraible no idea but it looks like an issue from master as every PR is failing with same error for gateways alone

**Comment by sendilkumarn** [comment]
@mraible I think it is broken because of the recent upgrade. I have tried to fix it in `jhipster-dependencies` by upgrading spring-cloud dependencies.

Before that gets merged you can escape this error by adding this dependency. 
`mavenBom 'org.springframework.cloud:spring-cloud-netflix:2.0.0.M6'` || override the complete spring-cloud dependency to `Finchley.M6`

**Comment by mraible** [comment]
Hello @jhipster/developers! I'm getting the following Travis error on this PR and I can't figure out how to fix it:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project travis-oauth-2-sass-infinispan: Compilation failure
[ERROR] /Users/mraible/dev/generator-jhipster/travis/samples/ngx-mariadb-oauth2-sass-infinispan-sample/src/main/java/io/github/jhipster/travis/service/BankAccountQueryService.java:[110,51] no suitable method found for buildReferringEntitySpecification(io.github.jhipster.service.filter.LongFilter,javax.persistence.metamodel.SingularAttribute<io.github.jhipster.travis.domain.BankAccount,io.github.jhipster.travis.domain.User>,javax.persistence.metamodel.SingularAttribute<io.github.jhipster.travis.domain.User,java.lang.String>)
[ERROR]     method io.github.jhipster.service.QueryService.<OTHER,X>buildReferringEntitySpecification(io.github.jhipster.service.filter.Filter<X>,javax.persistence.metamodel.SingularAttribute<? super io.github.jhipster.travis.domain.BankAccount,OTHER>,javax.persistence.metamodel.SingularAttribute<OTHER,X>) is not applicable
[ERROR]       (inferred type does not conform to equality constraint(s)
[ERROR]         inferred: java.lang.String
[ERROR]         equality constraints(s): java.lang.String,java.lang.Long)
[ERROR]     method io.github.jhipster.service.QueryService.<OTHER,X>buildReferringEntitySpecification(io.github.jhipster.service.filter.Filter<X>,javax.persistence.metamodel.SetAttribute<io.github.jhipster.travis.domain.BankAccount,OTHER>,javax.persistence.metamodel.SingularAttribute<OTHER,X>) is not applicable
[ERROR]       (cannot infer type-variable(s) OTHER,X
[ERROR]         (argument mismatch; javax.persistence.metamodel.SingularAttribute<io.github.jhipster.travis.domain.BankAccount,io.github.jhipster.travis.domain.User> cannot be converted to javax.persistence.metamodel.SetAttribute<io.github.jhipster.travis.domain.BankAccount,OTHER>))
```

I'm able to reproduce the issue locally and fix it, but I'm unsure of how to make this change in the entity generator. The change needed is to change `userId` in the generated criteria class from using a `LongFilter` to using a `StringFilter`. Any ideas are appreciated.

**Comment by mraible** [comment]
Nevermind, I figured it out. I had to change `getPkType()` in `generator-base-private.js`:

```js
/**
 * Decide the primary key type based on DB
 *
 * @param {any} databaseType - the database type
 */
getPkType(databaseType, authenticationType) {
    if (['cassandra', 'mongodb', 'couchbase'].includes(databaseType) || authenticationType === 'oauth2') {
        return 'String';
    }
    return 'Long';
}
```

**Comment by mraible** [comment]
Doh! That doesn't work because it changes all types. I just want to change the PK type for User.

**Comment by jdubois** [comment]
This is a very important feature for JHipster 5 -> please note that @pascalgrimaud is also doing https://github.com/jhipster/jhipster-registry/pull/244 and that this should all work together

**Comment by mraible** [comment]
Wahoo! Tests pass in Travis. @jhipster/developers can I get a review/approval please?

File: generators/server/files.js [hunk 1]
```diff
@@ -574,35 +574,31 @@ function writeFiles() {
                     this.template(`${SERVER_MAIN_SRC_DIR}package/domain/User.java.ejs`, `${javaDir}domain/User.java`);
                     this.template(`${SERVER_TEST_SRC_DIR}package/web/rest/AccountResourceIntTest.java.ejs`, `${testDir}web/rest/AccountResourceIntTest.java`);
                     this.template(`${SERVER_TEST_SRC_DIR}package/security/SecurityUtilsUnitTest.java.ejs`, `${testDir}security/SecurityUtilsUnitTest.java`);
+                    this.template(`${SERVER_MAIN_RES_DIR}config/liquibase/users.csv.ejs`, `${SERVER_MAIN_RES_DIR}config/liquibase/users.csv`);
+                    this.template(`${SERVER_MAIN_SRC_DIR}package/domain/Authority.java.ejs`, `${javaDir}domain/Authority.java`);
+                    this.template(`${SERVER_MAIN_SRC_DIR}package/service/UserService.java.ejs`, `${javaDir}service/UserService.java`);
+                    this.template(`${SERVER_MAIN_SRC_DIR}package/service/dto/package-info.java.ejs`, `${javaDir}service/dto/package-info.java`);
```
Comment (jdubois) [inline generators/server/files.js]: looks like you have a formatting issue here

**Comment by jdubois** [comment]
I love it because you remove much more code than you add :-)

File: generators/server/files.js (hunk 1 above)
Comment (mraible) [reply generators/server/files.js]: I don't see the issue you're referring to. Can you please be more specific?

Comment (jdubois) [reply generators/server/files.js]: I see some spaces at the beginning of the line - nothing important

**Comment by jdubois** [comment]
@mraible you know this very well, and we have a good test coverage here, so let's merge this

**Pull request merged**